File: Core/src/JavaScriptBridge/JSInjectorReporter.m
/*
 * This file is part of Adblock Plus <https://adblockplus.org/>,
 * Copyright (C) 2006-present eyeo GmbH
 *
 * Adblock Plus is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * Adblock Plus is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Adblock Plus.  If not, see <http://www.gnu.org/licenses/>.
 */

#import "JSInjectorReporter.h"
#import "UIBAlertView.h"
#import "Settings.h"
#import "Utils.h"
#import "SAContentWebView.h"
#import "UIBAlertView.h"
#import "NSJSONSerialization+NSString.h"
#import "ObjCLogger.h"

#import <KittCore/KittCore-Swift.h>

@interface JSInjectorReporter ()

@property (nonatomic, strong) NSArray *apiForContext;

@end

@implementation JSInjectorReporter

static NSString *API_SUCCESS_RETVAL = @"OK";
static NSString *API_SUCCESS_TOKEN = @"__SUCCESS__";

static NSString *ALL_FRAMES = @"_ALL_FRAMES_";
static NSString *ADDON_ID_TOKEN = @"%ADDON_ID%";
static NSString *TAB_ID_TOKEN = @"%TAB_ID%";
static NSString *BROWSERIFIED_API_TOKEN = @"__BROWSERIFIED_API__";
// Content script token must be a complete parseable JS entity, otherwise minification fails.
// This code has an advantage of doing an useful reporting when script injection fails
static NSString *USERSCRIPT_TOKEN = @"throw new Error('Content script not replaced')";
static NSString *CONTENT_SCRIPT_RUN_AT = @"%RUN_AT%";

// @warning Do not reorder unless apiForContext initialization is reordered too!
typedef NS_ENUM(NSUInteger, APIExecutionContext) {
    APIExecutionContext_BrowserAction,
    APIExecutionContext_Background,
    APIExecutionContext_ContentScript,
    APIExecutionContext_ContentDOM,
    APIExecutionContext_ContentMainframe,
    APIExecutionContext_ContentSubframe
};

/**
 @param apiFilename of the browserified API
 @param wrapperFilename of the injection context
 @param relevant execution wrapper as defined above
 */
- (NSString *)stringByInstrumentingBrowserifiedAPIFromJs:(NSString *)apiFilename bundle:(NSBundle *)bundle
{
    NSError *errFileLoad = nil;
    // load the prepared JS API template
    NSString *filePath = [bundle pathForResource:apiFilename ofType:@"js"];
    NSString *apiSource = [NSString stringWithContentsOfFile:filePath
                                                    encoding:NSUTF8StringEncoding
                                                       error:&errFileLoad];
    if (errFileLoad) {
        // Bundle construction problem, not a runtime problem. Crude reporting is good enough
        UIAlertView *alert = [Utils alertViewWithError:errFileLoad title:apiFilename delegate:nil];
        [alert show];
        return @""; // to prevent exception "insert nil" and allow the alert display
    }
    // replace tokens in wrapper
    apiSource = [apiSource stringByReplacingOccurrencesOfString:API_SUCCESS_TOKEN withString:API_SUCCESS_RETVAL];
    return apiSource;
}

- (instancetype)init
{
    return [self initWithBundle:[Settings coreBundle]];
}

- (instancetype)initWithBundle:(NSBundle *)bundle
{
    if (self = [super init]) {
        // prepare APIs just for %USERSCRIPT% replacement
        // @warning Follow the order of insertion defined by APIExecutionContext enum!
        _apiForContext =
            @[
               [self stringByInstrumentingBrowserifiedAPIFromJs:@"api_popup"
                                                         bundle:bundle],
               [self stringByInstrumentingBrowserifiedAPIFromJs:@"api_background"
                                                         bundle:bundle],
               [self stringByInstrumentingBrowserifiedAPIFromJs:@"api_content_script"
                                                         bundle:bundle],
               [self stringByInstrumentingBrowserifiedAPIFromJs:@"api_content_dom"
                                                         bundle:bundle],
               [self stringByInstrumentingBrowserifiedAPIFromJs:@"api_content_mainframe"
                                                         bundle:bundle],
               [self stringByInstrumentingBrowserifiedAPIFromJs:@"api_content_subframe"
                                                         bundle:bundle]
            ];
    }
    return self;
}

- (NSString *)stringWithBackgroundAPIForExtensionId:(NSString *)extensionId
{
    NSString *wrapperCopy = [NSString stringWithString:_apiForContext[APIExecutionContext_Background]];
    wrapperCopy = [wrapperCopy stringByReplacingOccurrencesOfString:ADDON_ID_TOKEN withString:extensionId];
    return wrapperCopy;
}

- (NSString *)stringWithBrowserActionAPIForExtensionId:(NSString *)extensionId
{
    NSString *wrapperCopy = [NSString stringWithString:_apiForContext[APIExecutionContext_BrowserAction]];
    wrapperCopy = [wrapperCopy stringByReplacingOccurrencesOfString:ADDON_ID_TOKEN withString:extensionId];
    return wrapperCopy;
}

- (NSString *)stringWithContentScriptAPIForExtensionId:(NSString *)extensionId
                                                 tabId:(NSUInteger)tabId
                                                 runAt:(NSString *)runAt
                                        wrappingScript:(NSString *)script
{
    NSString *tabIdStr = [@(tabId) stringValue];
    NSString *wrapperCopy = [NSString stringWithString:_apiForContext[APIExecutionContext_ContentScript]];
    wrapperCopy = [wrapperCopy stringByReplacingOccurrencesOfString:ADDON_ID_TOKEN withString:extensionId];
    wrapperCopy = [wrapperCopy stringByReplacingOccurrencesOfString:TAB_ID_TOKEN withString:tabIdStr];
    wrapperCopy = [wrapperCopy stringByReplacingOccurrencesOfString:USERSCRIPT_TOKEN withString:script];
    wrapperCopy = [wrapperCopy stringByReplacingOccurrencesOfString:CONTENT_SCRIPT_RUN_AT withString:runAt];
    return wrapperCopy;
}

- (NSString *)stringWithContentDOMAPIForExtensionId:(NSString *)extensionId
                                              tabId:(NSUInteger)tabId
{
    NSString *tabIdStr = [@(tabId) stringValue];
    NSString *wrapperCopy = [NSString stringWithString:_apiForContext[APIExecutionContext_ContentDOM]];
    wrapperCopy = [wrapperCopy stringByReplacingOccurrencesOfString:ADDON_ID_TOKEN withString:extensionId];
    wrapperCopy = [wrapperCopy stringByReplacingOccurrencesOfString:TAB_ID_TOKEN withString:tabIdStr];
    return wrapperCopy;
}

- (id)handleInjectionResult:(NSString *)evalResult
             withCallbackId:(NSString *)callbackId
                 withOrigin:(CallbackOriginType)origin
      errorReportProperties:(NSDictionary *)properties
              responseError:(NSError *__autoreleasing *)error
{

    NSString *callbackUndefRetval = [NSString stringWithFormat:@"%@ undefined", callbackId];

    if (evalResult && ![evalResult isKindOfClass:[NSString class]]) {
        return evalResult;
        // injection didn't succeed in some way
    } else if (!evalResult || [evalResult isEqualToString:@""]) {
        // should not happen, something is horribly wrong in the API (syntax etc.)
        evalResult = [NSString stringWithFormat:@"Callback %@ returned null/empty result", callbackId];
    } else if ([evalResult isEqualToString:callbackId]) {
        // callback executed correctly and didn't return any specific synchronous value
        // can return nil because the caller should know from command context that
        // return value is not expected. If there is error, responseError would be set.
        return nil;
    } else if ([evalResult hasPrefix:@"ERRORSTACKTRACE"]) {
        evalResult = [evalResult substringFromIndex:[@"ERRORSTACKTRACE" length]];
        evalResult = [self formattedParagraphFromErrorJSONString:evalResult];
    } else if ([evalResult isEqualToString:callbackUndefRetval]) {
        if (origin == CallbackOriginContent) {
            // content script has undefined callback wrapper, report but ignore
            LogWarn(@"Content callback wrapper does not exist, API not injected yet");
            return nil;
        }
    } else if ([evalResult hasPrefix:@"Callback id not found"]) {
        if (origin == CallbackOriginContent) {
            // content script has undefined callback, report but ignore
            LogWarn(@"JS callback '%@' does not exist, API not injected yet", callbackId);
            return nil;
        }
    } else {
        // should end up here with a callback return object
        id retval = [NSJSONSerialization JSONObjectWithString:evalResult options:0 error:error];
        if (*error) {
            evalResult = [NSString stringWithFormat:@"Expecting JSON parseable callback response, got '%@' error '%@'",
                                   evalResult, [*error localizedDescription]];
        } else {
            return retval;
        }
    }
    if (!(*error)) {
        *error = [NSError errorWithDomain:@"Kitt"
                                     code:0
                                 userInfo:@{ NSLocalizedDescriptionKey : evalResult }];
    }
    [self sendInjectionFailureReport:evalResult withExtraProperties:properties];
    return nil;
}

- (void)handleInjectionResult:(NSString *)evalResult
               withCallbackId:(NSString *)callbackId
                   withOrigin:(CallbackOriginType)origin
        errorReportProperties:(NSDictionary *)properties
                andCompletion:(CommandHandlerBackendCompletion)completion
{
    NSError *error;
    id result = [self handleInjectionResult:evalResult
                             withCallbackId:callbackId
                                 withOrigin:origin
                      errorReportProperties:properties
                              responseError:&error];
    completion(error, result);
}

- (BOOL)injectContentWindowGlobalSymbolsToWebView:(SAContentWebView *)webView
                            orNonMainFrameContext:(JSContext *)context
                                      isMainFrame:(BOOL)isMainFrame
                               scriptsInAllFrames:(BOOL)allFrames
{
    APIExecutionContext executionContext = isMainFrame ? APIExecutionContext_ContentMainframe : APIExecutionContext_ContentSubframe;
    NSString *tabIdStr = [@(webView.identifier) stringValue];
    NSString *wrapperCopy = [NSString stringWithString:_apiForContext[executionContext]];
    wrapperCopy = [wrapperCopy stringByReplacingOccurrencesOfString:ADDON_ID_TOKEN withString:kGlobalScopeExtId];
    wrapperCopy = [wrapperCopy stringByReplacingOccurrencesOfString:TAB_ID_TOKEN withString:tabIdStr];
    wrapperCopy = [wrapperCopy stringByReplacingOccurrencesOfString:ALL_FRAMES withString:allFrames ? @"T" : @"F"];
    NSString *urlString = webView.currentURL.absoluteString;
    return [self injectJavaScriptCode:wrapperCopy
                            toWebView:webView
                            orContext:isMainFrame ? nil : context
                errorReportProperties:@{
                    @"context" : @"content window global symbols",
                    @"url" : urlString ? urlString : [NSNull null]
                }];
}

- (BOOL)injectJavaScriptCode:(NSString *)jsCode
                   toWebView:(UIWebView *)webView
                   orContext:(JSContext *)context
       errorReportProperties:(NSDictionary *)properties
{
    // JS evaluation in UIWebView can only return string
    // The JS API is instrumented to return a specific string on success,
    // or a JSON stringified Error on failure
    NSString *evalResult;
    if (context != nil) {
        JSValue *value = [context evaluateScript:jsCode];
        if (context.exception) {
            value = context.exception;
        }
        evalResult = [value toString];
    } else {
        evalResult = [webView stringByEvaluatingJavaScriptFromString:jsCode];
    }
    if ([API_SUCCESS_RETVAL isEqualToString:evalResult]) {
        return YES;
    }
    // injection didn't succeed in some way
    if (!evalResult || [evalResult isEqualToString:@""]) {
        // should not happen, something is horribly wrong in the API (syntax etc.)
        evalResult = @"Eval returned null/empty result";
    } else {
        evalResult = [self formattedParagraphFromErrorJSONString:evalResult];
    }
    [self sendInjectionFailureReport:evalResult withExtraProperties:properties];
    return NO;
}

- (void)sendInjectionFailureReport:(NSString *)report
               withExtraProperties:(NSDictionary *)properties
{
    // not a dev mode, log silently to service
    NSMutableDictionary *props = [NSMutableDictionary dictionaryWithDictionary:properties];
    props[@"report"] = report;
    LogError(@"INJECTION FAILED\n%@\n%@", report, properties);
}

- (NSString *)formattedParagraphFromErrorJSONString:(NSString *)jsonString
{
    // Unpack the returned string as stack trace
    NSError *parsingError = nil;
    id errorDictionary = [NSJSONSerialization JSONObjectWithString:jsonString options:0 error:&parsingError];
    if (parsingError) {
        return [NSString stringWithFormat:@"Eval result JSON is invalid:%@\n%@",
                         [parsingError localizedDescription], jsonString];
    } else if (![errorDictionary isKindOfClass:[NSDictionary class]]) {
        return [NSString stringWithFormat:@"Eval result is not NSDictionary but %@:\n%@",
                         NSStringFromClass([errorDictionary class]), jsonString];
    } else if ([errorDictionary count] == 0) {
        return [NSString stringWithFormat:@"Eval result is an empty dictionary:\n%@", jsonString];
    }
    NSString *parsedRetval = @"";
    // errorDictionary is a JSONized JS Error object: all properties
    // which JavaScriptCore has put in it.
    for (NSString *frameKey in errorDictionary) {
        id frameValue = errorDictionary[frameKey];
        bool isFrameKeyParsed = NO; // was the current frame key already parsed?
        /*
     Output lines are formatted as "key : value" per key, with exception of any
     value which contains a multiline string. Most prominently the "stack" key.
     Remember this is what JavaScriptCore gave us, not something we created
     ourselves. Such value will be splitted and formatted to multiple
     "key : value" lines.
    */
        // need to check the value type. JSONKit may have given us a different type,
        // like NSNumber or NSNull.
        if ([frameValue isKindOfClass:[NSString class]]) {
            NSArray *frameValueMultiline = [frameValue componentsSeparatedByString:@"\n"];
            if ([frameValueMultiline count] > 1) {
                NSUInteger lineNumber = 0;
                for (NSString *line in frameValueMultiline) {
                    parsedRetval = [parsedRetval stringByAppendingFormat:@"%@ #%lu: %@\n",
                                                 frameKey, (unsigned long)++lineNumber, line];
                }
                isFrameKeyParsed = YES;
            }
        }
        // the above multiline check didn't kick in
        // fall back to default single line formatting
        if (!isFrameKeyParsed) {
            parsedRetval = [parsedRetval stringByAppendingFormat:@"%@ : %@\n",
                                         frameKey, frameValue];
        }
    }
    return parsedRetval;
}
@end

File: Core/src/JavaScriptBridge/JSInjectorReporter.swift
/*
 * This file is part of Adblock Plus <https://adblockplus.org/>,
 * Copyright (C) 2006-present eyeo GmbH
 *
 * Adblock Plus is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * Adblock Plus is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Adblock Plus.  If not, see <http://www.gnu.org/licenses/>.
 */

import Foundation

extension JSInjectorReporter {
    private static let callbackObjectName = "KittCallbackCaller"
    private static let callbackFunctionName = "invoke"

    private func createObject(from javaScriptCallback: BridgeCallback, with payload: Any?) -> ([String: Any]) {
        var context = javaScriptCallback.context
        let object: [String: Any]
        if let error = payload as? NSError {
            context["lastError"] = ["message": error.localizedDescription]
            object = [
                "context": context
            ]
        } else {
            object = [
                "context": context,
                "data": payload as Any
            ]
        }
        return object
    }

    final func call(_ javaScriptCallback: BridgeCallback, with payload: Any?, completion: StandardCompletion?) {
        assert(Thread.isMainThread, "Must be run on main thread!")

        switch javaScriptCallback.webView {
        case .some(let webView as SAWebView):
            call(javaScriptCallback, with: payload, from: webView, completion: completion)
        case .some(let webView as BackgroundWebView):
            call(javaScriptCallback, with: payload, from: webView, completion: completion)
            /*case .Some(let webView as ContentWebView):
             call(javaScriptCallback, fromWebView: webView, completion: completion)
             break*/
        default:
            completion?(.failure(NSError(message: "WebView is not set")))
        }
    }

    private func call(_ javaScriptCallback: BridgeCallback,
                      with payload: Any?,
                      from webView: WebViewFacade,
                      completion: StandardCompletion?) {
        let object = createObject(from: javaScriptCallback, with: payload)
        let origin = javaScriptCallback.origin
        let extensionId = javaScriptCallback.extension?.extensionId
        let callbackId = javaScriptCallback.callbackId

        let properties: [String: Any] = [
            "context": "callback \(Utils.callbackOriginDescription(origin))",
            "extension": extensionId as Any,
            "parameters": object,
            "url": webView.url?.absoluteString as Any
        ]
        // swiftlint:disable:next force_try
        let json = try! JSONSerialization.string(withJSONObject: object, options: JSONSerialization.WritingOptions())
        let javascript = "window.\(JSInjectorReporter.callbackObjectName).\(JSInjectorReporter.callbackFunctionName)(\(json))"
        webView.evaluateJavaScript(javascript) { [weak self, weak webView] result, error in
            if let uwError = error {
                if let context = (webView as? BackgroundWebView)?.context, let extensionId = extensionId {
                    if shouldReloadBackground(for: uwError) {
                        context.reloadBackground(for: extensionId)
                    }
                }
                completion?(.failure(uwError))
            } else if let injector = self, let uwResult = result as? String {
                injector.handleInjectionResult(uwResult,
                                               withCallbackId: callbackId,
                                               with: origin,
                                               errorReportProperties: properties) { error, result in
                    if let uwError = error {
                        completion?(.failure(uwError))
                    } else {
                        completion?(.success(result))
                    }
                }
            } else {
                completion?(.failure(NSError(message: "Injector has not been found")))
            }
        }
    }

    private func call(_ javaScriptCallback: BridgeCallback,
                      with payload: Any?,
                      from webView: SAWebView,
                      completion: StandardCompletion?) {
        let callbackId = javaScriptCallback.callbackId
        guard let webThread = self.webThread else {
            completion?(.failure(NSError(message: "webThread is not set")))
            return
        }

        guard let frame = javaScriptCallback.frame else {
            let error = "Callback \(callbackId) KittFrame not found for WebKit frame \(String(describing: javaScriptCallback.frame))"
            completion?(.failure(NSError(message: error)))
            return
        }

        guard let kittFrame = webView.kittFrame(forWebKitFrame: frame), let jsContext = kittFrame.context else {
            let error = "Callback \(callbackId) KittFrame found for WebKit frame but has no JS context"
            completion?(.failure(NSError(message: error)))
            return
        }

        let object = createObject(from: javaScriptCallback, with: payload)
        let origin = javaScriptCallback.origin
        let properties: [String: Any] = [
            "context": "callback \(Utils.callbackOriginDescription(origin))",
            "extension": javaScriptCallback.extension?.extensionId as Any,
            "parameters": object,
            "url": webView.url?.absoluteString as Any
        ]

        // This completion block is called from Javascript.
        // Response contains either result or error.
        let jsCompletion: JSCompletion = { [weak self] (response) in
            // Dispatch from web thread to main thread
            DispatchQueue.main.async { [weak self] () -> Void in
                switch response {
                case .failure(let error):
                    completion?(.failure(error))
                case .success(let success):
                    if let injector = self {
                        injector.handleInjectionResult(success,
                                                       withCallbackId: callbackId,
                                                       with: origin,
                                                       errorReportProperties: properties) { error, result in
                            if let uwError = error {
                                completion?(.failure(uwError))
                            } else {
                                completion?(.success(result))
                            }
                        }
                    } else {
                        completion?(.failure(NSError(message: "Injector has not been found")))
                    }
                }
            }
        }

        // This dispatch to web thread is needed, otherwise callbacks are too unstable.
        let callbackContext = CallbackContext(context: jsContext, object: object, callbackId: callbackId, completion: jsCompletion)
        perform(#selector(callJavaScriptCallbackWith), on: webThread, with: callbackContext, waitUntilDone: false)
    }

    @objc
    public final func callJavaScriptCallback(_ callback: BridgeCallback, completion: CommandHandlerBackendCompletion?) {
        call(callback, with: nil) { result in
            switch result {
            case .success(let output):
                completion?(nil, output)
            case .failure(let error as IgnorableError):
                if completion == nil {
                    Log.info("Callback execution: \(error.localizedDescription)")
                }
                completion?(error, nil)
            case .failure(let error):
                if completion == nil {
                    Log.error("Callback execution: \(error.localizedDescription)")
                }
                completion?(error, nil)
            }
        }
    }

    // MARK: - Private

    private typealias JSCompletion = (Result<String>) -> Void

    private class CallbackContext: NSObject {
        let context: JSContext
        let object: Any
        let callbackId: String
        let completion: JSCompletion

        init(context: JSContext, object: Any, callbackId: String, completion: @escaping JSCompletion) {
            self.context = context
            self.object = object
            self.callbackId = callbackId
            self.completion = completion
            super.init()
        }
    }

    @objc
    private final func callJavaScriptCallbackWith(_ callbackContext: CallbackContext) {
        let context = callbackContext.context
        let completion = callbackContext.completion

        guard let window = context.globalObject,
            let callbackObject = window.objectForKeyedSubscript(JSInjectorReporter.callbackObjectName),
            !callbackObject.isUndefined && !callbackObject.isNull else {
                completion(.failure(NSError(message: "Callback \(callbackContext.callbackId) JS window does not contain the entry symbol")))
                return
        }

        guard let callbackFunction = callbackObject.objectForKeyedSubscript(JSInjectorReporter.callbackFunctionName),
            !callbackFunction.isUndefined && !callbackFunction.isNull else {
                completion(.failure(NSError(message: "JSC callback \(callbackContext.callbackId) fn not found")))
                return
        }

        let optionalResult = callbackFunction.call(withArguments: [callbackContext.object])

        guard let result = optionalResult, !result.isUndefined && !result.isNull, let string = result.toString() else {
            completion(.failure(NSError(message: "JSC callback must have a retval")))
            return
        }

        completion(.success(string))
    }
}

private func shouldReloadBackground(`for` error: Error) -> Bool {
    let codes = [WKError.unknown.rawValue,
                 WKError.webContentProcessTerminated.rawValue,
                 WKError.webViewInvalidated.rawValue]
    let error = error as NSError
    return error.domain == WKErrorDomain && codes.contains(error.code)
}
